File: linked_list/86.partition-list.cpp
/*
 * @lc app=leetcode id=86 lang=cpp
 *
 * [86] Partition List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode* partition(ListNode* head, int x) {
    ListNode first_head, *first_tail = &first_head;
    ListNode second_head, *second_tail = &second_head;

    while (head) {
      if (head->val < x) {
        first_tail->next = head;
        first_tail = head;
      } else {
        second_tail->next = head;
        second_tail = head;
      }
      head = head->next;
    }

    first_tail->next = second_head.next;
    second_tail->next = nullptr;

    return first_head.next;
  }
};
// @lc code=end

File: 86.partition-list.py
#
# @lc app=leetcode id=86 lang=python3
#
# [86] Partition List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        left_head, right_head = ListNode(), ListNode()
        left_tail, right_tail = left_head, right_head

        while head:
            if head.val < x:
                left_tail.next = head
                left_tail = head
            else:
                right_tail.next = head
                right_tail = head
            head = head.next

        left_tail.next = right_head.next
        if right_tail:
            right_tail.next = None

        return left_head.next


# @lc code=end
